File: js/test/PeekIterator.test.js
const PeekIterator = require("../common/PeekIterator")
const arrayToGenerator = require("../common/arrayToGenerator")

const {assert } = require("chai")

describe('test PeekIterator',()=>{
    it('test_peek',  ()=> {
        // [..."abcdefg"]   字符串转字节数组
        const iter = new PeekIterator(arrayToGenerator([..."abcdefg"]))
        assert.equal(iter.next(),'a')
        assert.equal(iter.next(),'b')
        assert.equal(iter.peek(),'c')
        assert.equal(iter.peek(),'c')
        assert.equal(iter.next(),'c')
        assert.equal(iter.next(),'d')
    })

    it('test_lookahead2', ()=> {
        const iter = new PeekIterator(arrayToGenerator([..."abcdefg"]))
        assert.equal(iter.next(),'a')
        assert.equal(iter.peek(),'b')
        assert.equal(iter.peek(),'b')
        assert.equal(iter.next(),'b')
        assert.equal(iter.next(),'c')
        iter.putBack()
        iter.putBack()
        assert.equal(iter.peek(),'b')
        assert.equal(iter.peek(),'b')
        assert.equal(iter.next(),'b')
        assert.equal(iter.next(),'c')
    });

    it('test_endToken',  ()=> {
        let source = "abcdefg";
        const iter = new PeekIterator(arrayToGenerator([...source]))

        for (let i = 0; i < source.length+100; i++) {
            if (i===7){
                assert.equal(iter.next(),'\0')
            }else{
                assert.equal(iter.next(),source[i])
            }
        }
    });
})

File: src/test/java/common/PeekIteratorTest.java
package common;


import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class PeekIteratorTest {

    @Test
    public void test_peak() {
        String source = "abcdefg";
        PeekIterator<Character> it = new PeekIterator<>(source.chars().mapToObj(c -> (char) c), (char) 0);

        assertEquals('a', it.next().charValue());
        assertEquals('b', it.next().charValue());

        it.next();
        it.next();

        assertEquals('e', it.next().charValue());
        assertEquals('f', it.peek().charValue());
        assertEquals('f', it.peek().charValue());
        assertEquals('f', it.next().charValue());
        assertEquals('g', it.next().charValue());
    }
    @Test
    public  void test_lookahead2(){
        String source = "abcdefg";
        PeekIterator<Character> it = new PeekIterator<>(source.chars().mapToObj(c -> (char) c),(char)0);

        assertEquals('a',it.next().charValue());
        assertEquals('b',it.next().charValue());
        assertEquals('c',it.next().charValue());
        it.putBack();
        it.putBack();
        assertEquals('b',it.next().charValue());
    }

    @Test
    public void test_endToken(){
        String source = "abcdefg";
        PeekIterator<Character> it = new PeekIterator<>(source.chars().mapToObj(c -> (char) c), (char) 0);
        int i =0;
        while(it.hasNext()){
            if (i==source.length()){
                assertEquals((char)0,it.next().charValue());
            }else{
                assertEquals(source.charAt(i),it.next().charValue());
                i++;
            }
        }
    }
}